### ISSUE TITLE
TypeError: register() missing 1 required positional argument: 'constraint_value'

### ISSUE THREAD

**Issue by xmaww** [fmfn/BayesianOptimization#402]
`# -*- coding: utf-8 -*-
"""
Created on Wed Mar 15 10:41:32 2023

@author: lmy818
"""

import numpy as np
from bayes_opt import BayesianOptimization
import matplotlib.pyplot as plt
from scipy.optimize import NonlinearConstraint
from bayes_opt import UtilityFunction
utility = UtilityFunction(kind="ucb", kappa=500, xi=0.0)

# def target_function(x, y):
#     # Gardner is looking for the minimum, but this packages looks for maxima, thus the sign switch
#     return np.cos(2*x)*np.cos(y) + np.sin(x)

def constraint_function(x, y):
    return x+y
constraint_limit_h = 18.33
constraint_limit_l = 0.0
constraint = NonlinearConstraint(constraint_function, constraint_limit_l, constraint_limit_h)
# Bounded region of parameter space
pbounds = {'x': (0, 18.33), 'y': (0, 18.33)}

optimizer = BayesianOptimization(
    f=None,
    constraint=constraint,
    pbounds=pbounds,
    verbose=2, # verbose = 1 prints only when a maximum is observed, verbose = 0 is silent
    random_state=4,
)
optimizer.register(
    params={'x': 1.33, 'y': 8.33},
    target=813.2151776,
    # constraint_value=0, help!!!
    )
next_point_to_probe = optimizer.suggest(utility)`
Here is my code aiming to deal with a 2-parameter and 1-linear-constraint problem，the error occur when excuting the line**“optimizer.register(
    params={'x': 1.33, 'y': 8.33},
    target=813.2151776,
    # constraint_value=0, help!!!
    )”**
and when I look into the source code of the class BayesianOptimization，I can‘t find a parameter called 'constraint_value',is there anyone who meet the same problem before?How deal with it?Many thanks!!!

**Comment by till-m** [comment]
1. If you want to perform constrained optimization, you need to call register with the value of the constraint function at the point you're registering.
2. Are you sure you're version is 1.4.1? If I run your code I get a different error message: `ValueError: When registering a point to a constrained TargetSpace a constraint value needs to be present.`

**Comment by xmaww** [comment]
Many thanks！
I updated my version and it can run without bug!And I found the version I called is 1.3.0,you are right.

**Comment by xmaww** [comment]
Currently,after I added one constraint_function,I found that the BayesianOptimization.suggest() returns improper next_time_to_probe point.The code is as follows:
```python
import numpy as np
# fixed (xb,yb) at (9.165,13.71),yc at 22.85
from bayes_opt import BayesianOptimization,UtilityFunction
# import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm
from matplotlib import mlab
from matplotlib import gridspec
from matplotlib.backends.backend_pdf import PdfPages
# import matplotlib.pyplot as plt
from scipy.optimize import NonlinearConstraint
utility = UtilityFunction(kind="ucb", kappa=10, xi=0.0)

def constraint_function(yA, yB,yC):
    if yC-yB>0 and yC-yA>0:
        return 1
    return 0  
# constraint_function(4,2,3)
constraint =  NonlinearConstraint(constraint_function, 0.5, np.inf)
bo = BayesianOptimization(
    f=None,
    constraint=constraint,
    pbounds={'xA':(0,18.33), 'yA':(0, 18.28),'xB':(0,18.33), 'yB':(0, 18.28),'yC':(9.14,27.42)},
    verbose=0, # verbose = 1 prints only when a maximum is observed, verbose = 0 is silent
    random_state=2,
)
#1
bo.register(
    params={'xA':17.7257,'xB':10.0308,'yA':17.7807,'yB':13.0668,'yC':21.8945},
    target=-1.2197645,
    constraint_value = 1
)

#2
bo.register(
    params={'xA':6.0550,'xB':3.7512,'yA':11.3203,'yB':5.4777,'yC':20.7188},
    target=-2.5477375,
    constraint_value = 1
)

bo.register(
    params={'xA':6.1505,'xB':8.5470,'yA':2.1459,'yB':10.6810,'yC':26.5324},
    target=-3.25197125,
    constraint_value = 1
)
# next_point_to_probe = bo.suggest(utility)
# print("Next point to probe is:", next_point_to_probe)
# bo.register(
#     params={'xA':18.33,'xB':17.348860052646803,'yA':18.28,'yB':18.28,'yC':26.5324},
#     target=-3.25197125,
#     constraint_value = 1
# )
next_point_to_probe = bo.suggest(utility)
print("Next point to probe is:", next_point_to_probe)
# plot_2d()
```
The output is:
"Next point to probe is: {'xA': 18.33, 'xB': 15.756315441427272, 'yA': 18.28, 'yB': 18.28, 'yC': 16.076954429069918}",
but I have the constraint to make 'yC' larger that 'yB' and 'yA',but the output is not that correct.Is there any problem when I define the constraint function?

**Comment by xmaww** [comment]
Actually in my problem,if the suggested point is out of the constraint,I have no way to get the function_value  because of it is actually a physical experiment process.Is there any solutions to make the parameters obtained by bo.suggest(utility) within constraint?

**Comment by till-m** [comment]
Hi @xmaww

your problem is interesting. I see a few ways to proceed, I will state them below. First of all, I have updated your code:

```python
import numpy as np
# fixed (xb,yb) at (9.165,13.71),yc at 22.85
from bayes_opt import BayesianOptimization,UtilityFunction
# import numpy as np
import matplotlib.pyplot as plt
from matplotlib import cm
from matplotlib import mlab
from matplotlib import gridspec
from matplotlib.backends.backend_pdf import PdfPages
# import matplotlib.pyplot as plt
from scipy.optimize import NonlinearConstraint
utility = UtilityFunction(kind="ucb", kappa=10, xi=0.0)

# The constraint here is that yC >= yA & yC>= yB.
# There are a few ways of modelling this constraint so that BO can use it.
# Remember, BO expects a function c(yA, yB, yC) ∈ ℝ and limits u, l.
# The constraint is fulfilled if l <= c(yA, yB, yC) <= u.
# One possible way of defining this constraint (considering that yA, yB, yC >=0)
def constraint_function(yA, yB, yC): # c(yA, yB, yC)
    return np.array([yC - yA, yC - yB])

# Formalizing the constraint, i.e. that 0.0 <= c(yA, yB, yC) (<= ∞)
constraint =  NonlinearConstraint(constraint_function, np.array([0.0, 0.0]), np.array([np.inf, np.inf]))

bo = BayesianOptimization(
    f=None,
    constraint=constraint,
    pbounds={'xA':(0,18.33), 'yA':(0, 18.28),'xB':(0,18.33), 'yB':(0, 18.28),'yC':(9.14,27.42)},
    verbose=0, # verbose = 1 prints only when a maximum is observed, verbose = 0 is silent
    random_state=2,
)

initial_params = [
    {'xA':17.7257,'xB':10.0308,'yA':17.7807,'yB':13.0668,'yC':21.8945},
    {'xA':6.0550,'xB':3.7512,'yA':11.3203,'yB':5.4777,'yC':20.7188},
    {'xA':6.1505,'xB':8.5470,'yA':2.1459,'yB':10.6810,'yC':26.5324}
]
intial_target_values = [-1.2197645, -2.5477375, -3.25197125]

for params, target_value in zip(initial_params, intial_target_values):
    print(constraint_function(params['yA'], params['yB'], params['yC']))
    bo.register(
        params=params,
        target=-1.2197645,
        constraint_value = constraint_function(params['yA'], params['yB'], params['yC'])
    )
    # NB: constraint_value is _not_ a boolean of whether the function is fulfilled, but
    # the actual value of the constraint function. Internally, we consider the likelihood
    # of the constraint being fulfilled, so it is important to know the actual value here.

next_point_to_probe = bo.suggest(utility)
print("Next point to probe is:", next_point_to_probe)
print(next_point_to_probe['yC'] - next_point_to_probe['yA'])
print(next_point_to_probe['yC'] - next_point_to_probe['yB'])

#Next point to probe is: {'xA': 9.849359522783159, 'xB': 4.457409792645482, 'yA': 0.3402968196793676, 'yB': 4.829759743770428, 'yC': 11.289400424349601}
#10.949103604670233
#6.459640680579174
```

Now, the constrained optimization in this package is designed for efficient optimiziation under constraint functions which are hard to evaluate, (specifically, roughly as hard as the objective function), but not constraints that are physically impossible to realize/and constraints that are easy to check.
In particular, this makes it hard to know a priori whether a specific point fulfills the constraint. Thus we model the likelihood of the constraint being fulfilled and make use of this information when suggesting new points. We will then evaluate the target function and the constraint. This will

1.  give us more information about the target function to use and allow us to construct a more accurate model of it.
2. more information about the constraint, which will help us make more accurate predictions about whether it is fulfilled or not. 

In particular, the idea is to make use of points that are technically not acceptable as solutions (since they don't fulfill the constraints) by considering the information we get from them about the target function, since a priori we do not know whether the point fulfills the constraints, since evaluating the constraint function is hard.

Now, in your problem the situation is different:
1. Your constraint function is _easy_ to check
2. It is _impossible_ to evaluate the the target function outside of the constraints.

As for your options, here's a few ideas:
1. Return a bad value whenever the optimizer suggests a point outside of the constraints:
```python
def my_target_function(params):
    return np.sum([params[key] for key in params])

def my_surrogate_function(params):
    bad_value = -10. #adjust for your specific problem
    if params['yC'] < params['yA'] or params['yC'] < params['yC']:
         return bad_value
    return my_target_function(params)
```
2. Optimize an unconstrained surrogate function in which you enforce the restriction through clamping:
```python
def my_target_function(params):
    return np.sum([params[key] for key in params])

def my_surrogate_function(params):
    params['yC'] = np.max((params['yC'], params['yB'], params['yA']))
    return my_target_function(params)
```
3. If you don't care too much about the computational overhead induced by the approximation of the constraint function, you could simply run the code above as-is. If a point is suggested that doesn't fulfil your constraint function, you can clamp it to the acceptable region by by using `yC = np.min((yA, yB))`. Since you're already using the suggest-evaluate-register protocol, this should be a good option. Since you're continually providing more and more information on the constrained region, the optimizer should not suggest too many points there.
```python
def clamp_if_needed(params):
    params['yC'] = np.max((params['yC'], params['yB'], params['yA']))
    return params['yC']

next_point_to_probe = clamp_if_needed(bo.suggest(utility))
```

The difference between options 2. & 3. is, that in option 2. you register the point under the params suggested by the optimizer and don't use a constraint function, whereas in option 3. you use a constraint function to improve suggestions + clamp 'yC' to the acceptable region and register the point there.

This is just off the top of my head -- there might be better ways of dealing with your problem. I hope this helps you get started.

**Comment by xmaww** [comment]
Excellent advice! I 'll give the above methods a try, thanks!
